### PR TITLE
Dev v1.2.0

### PR DESCRIPTION
**Pull request by gschmidt-usgs**
Peer review for dev_v1.2.0 branch

### CODE REVIEW

File: pixel_qa/generate_pixel_qa.c [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+    {
+        sprintf (errmsg, "Unable to read the entire Level-1 QA band");
+        error_handler (true, FUNC_NAME, errmsg);
+        return (ERROR);
+    }
+
+    /* Close the Level-1 QA file */
+    close_level1_qa (l1_fp_bqa);
+
+    /* Allocate memory for the pixel QA band; also allocates these values
```
Comment (raydittmeier) [inline pixel_qa/generate_pixel_qa.c]: Should "allocates" be "initializes"?

Comment (raydittmeier) [inline pixel_qa/generate_pixel_qa.c]: Should "allocates" be "initializes"?

Comment (raydittmeier) [inline pixel_qa/generate_pixel_qa.c]: Should "allocates" be "initializes"?

File: pixel_qa/generate_pixel_qa.c [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+
+    /* Loop through the pixels in the Level-1 QA band and create the pixel QA
+       band.  Buffers were already initialized with the L2QA_CLEAR bit set.  It
+       needs to be turned off if any condition is turned on, except the low or
+       moderate cloud confidence. */
+    /* Water is not available in the Level-1 QA.  The snow class is based on
+       the snow/ice confidence, and the cloud shadow class is based on the 
+       cloud shadow confidence.  The pixel QA will be turned on for snow and
+       cloud shadow if the confidence is high (i.e. both bits turned on is a
+       value of 3). */
```
Comment (raydittmeier) [inline pixel_qa/generate_pixel_qa.c]: For ease of reading, maybe make an enumerated type for 3, 2, and 1 when those values are used for high, moderate, or low confidence (like in the code below).

File: pixel_qa/read_pixel_qa.c [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+RETURN VALUE:
+Type = FILE *
+Value           Description
+-----           -----------
+NULL            Error parsing the XML file or opening the pixel QA band
+not NULL        Successfully read
+
+NOTES:
+1. It is expected that this QA band will be an unsigned 8-bit integer. If the
```
Comment (raydittmeier) [inline pixel_qa/read_pixel_qa.c]: Update the number of bits.  (I think it's referring to the new 16-bit QA band).

File: pixel_qa/read_pixel_qa.h [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+);
+
+int read_pixel_qa
+(
+    FILE *fp_bqa,           /* I: pointer to pixel QA band open for reading */
+    int nlines,             /* I: number of lines to read from the QA file */
+    int nsamps,             /* I: number of samples to read from the QA file */
+    uint16_t *pixel_qa      /* O: pixel QA band values for the specified number
+                                  of lines (memory should be allocated for
+                                  nlines x nsamps of size uint8 before calling
```
Comment (raydittmeier) [inline pixel_qa/read_pixel_qa.h]: Update the number of bits in the comment.

File: tools/test_read_pixel_qa.c [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+PROJECT:  Land Satellites Data System Science Research and Development (LSRD)
+at the USGS EROS
+
+LICENSE TYPE:  NASA Open Source Agreement Version 1.3
+
+NOTES:
+  1. The XML metadata format parsed or written via this library follows the
+     ESPA internal metadata format found in ESPA Raw Binary Format v1.0.doc.
+     The schema for the ESPA internal metadata format is available at
+     http://espa.cr.usgs.gov/schema/espa_internal_metadata_v1_0.xsd.
```
Comment (raydittmeier) [inline tools/test_read_pixel_qa.c]: I think I saw generic version numbers for the metadata file being used earlier.

File: pixel_qa/generate_pixel_qa.c (hunk 1 above)
Comment (gschmidt-usgs) [inline pixel_qa/generate_pixel_qa.c]: Fixed

File: pixel_qa/generate_pixel_qa.c (hunk 2 above)
Comment (gschmidt-usgs) [inline pixel_qa/generate_pixel_qa.c]: Done

File: pixel_qa/pixel_qa_dilation.c [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+#ifdef _OPENMP
+    #pragma omp parallel for private(start_row, end_row, row_index, col, output_index, start_col, end_col, found, window_row, window_row_index, window_col, input_index)
+#endif
+    for (row = 0; row < nrows; row++)
+    {
+        start_row = row - distance;
+        end_row = row + distance;
+
+        row_index = row * ncols;
+
```
Comment (gschmidt-usgs) [inline pixel_qa/pixel_qa_dilation.c]: Remove row_index and set output_index here.
output_index = row * ncols;

Increment output_index by one in the for loop.
for (col = 0; col < ncols; col++, output_index++)

If you choose to do this, you can remove the row_index from the parallel pragma as well.

File: pixel_qa/pixel_qa_dilation.c [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+                continue;
+            }
+
+            start_col = col - distance;
+            end_col = col + distance;
+
+            found = false;
+            /* For each row in the window */
+            for (window_row = start_row;
+                 window_row < end_row + 1 && !found;
```
Comment (gschmidt-usgs) [inline pixel_qa/pixel_qa_dilation.c]: Consider using
window_row <= end_row            versus
window_row < end_row + 1

This removes the +1 computation in the loop.  It probably gets optimized out, but might make it easier to read (especially with the && !found.  If you choose to do this, the same applies for the window_col loop below.

File: pixel_qa/pixel_qa_dilation.c [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    if (search_bit == L2QA_CLOUD)
+    {
+        cleaning_bit_mask &= ~(1 << L2QA_CLEAR);
+        cleaning_bit_mask &= ~(1 << L2QA_CLD_SHADOW);
+    }
+
+#ifdef _OPENMP
+    #pragma omp parallel for private(start_row, end_row, row_index, col, output_index, start_col, end_col, found, window_row, window_row_index, window_col, input_index)
+#endif
+    for (row = 0; row < nrows; row++)
```
Comment (gschmidt-usgs) [inline pixel_qa/pixel_qa_dilation.c]: It appears this algorithm basically loops through every pixel in the image, skipping fill pixels.  Then, for each pixel, the specified dilation window is searched to see if any of the surrounding window pixels have the desired bit "turned on".  As soon as a window pixel is found, then the loop is exited and the current pixel is "turned on"/dilated.

File: pixel_qa/pixel_qa_dilation.c [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    int cleaning_bit_mask; /* Mask based on bits to clean because otherwise 
+                           they would contradict the search bit */
+
+    /* Set the mask to dilate the user-selected bit. */
+    user_bit_mask = pow(2, search_bit); 
+
+    /* Initialize all bits in the cleaning mask to 1 */ 
+    cleaning_bit_mask = ~(cleaning_bit_mask & 0);
+
+    /* If cloud is being dilated, then turn off clear and cloud shadow.  Leave
```
Comment (gschmidt-usgs) [inline pixel_qa/pixel_qa_dilation.c]: This ends up being a nice solution for handling clean up (turn off) of other QA bits when the clouds are dilated.

File: pixel_qa/read_pixel_qa.c (hunk 3 above)
Comment (gschmidt-usgs) [inline pixel_qa/read_pixel_qa.c]: fixed

File: pixel_qa/read_pixel_qa.h (hunk 4 above)
Comment (gschmidt-usgs) [inline pixel_qa/read_pixel_qa.h]: fixed

File: tools/dilate_pixel_qa.c [hunk 10]
```diff
@@ -84,7 +85,7 @@ short get_args
     int argc,              /* I: number of cmd-line args */
     char *argv[],          /* I: string of cmd-line args */
     char **xml_infile,     /* O: address of input XML filename */
-    uint8_t *class_value,  /* O: address of class value variable*/
+    uint8_t *bit_value,    /* O: address of bit value variable */
     uint8_t *distance      /* O: address of distance value variable */
```
Comment (gschmidt-usgs) [inline tools/dilate_pixel_qa.c]: Actually, both bit_value and distance are I/O: arguments versus O:  Looking at the code, get_args expects that both these variables are initialized to 255 before calling get_args.  Might consider having the bit_value and distance initialized to 255 at the very beginning of get_args (before looping through the command-line options versus expecting them to be initialized to 255 in the main routine.  If not, at least document the expectation.

File: tools/test_read_pixel_qa.c [hunk 11]
```diff
@@ -0,0 +1,300 @@
...
+
+RETURN VALUE:
+Type = None
+
+NOTES:
+******************************************************************************/
+void usage ()
+{
+    printf ("test_read_pixel_qa is a simple test program that opens, reads, "
+            "and closed the Level-2 pixel QA band. It prints out a few pixel "
```
Comment (gschmidt-usgs) [inline tools/test_read_pixel_qa.c]: fixed

**Comment by gschmidt-usgs** [comment]
Done with comments on this peer review

File: tools/dilate_pixel_qa.c (hunk 10 above)
Comment (raydittmeier) [inline tools/dilate_pixel_qa.c]: fixed by moving the initialization so they aren't I/O

File: pixel_qa/pixel_qa_dilation.c (hunk 8 above)
Comment (raydittmeier) [inline pixel_qa/pixel_qa_dilation.c]: fixed with a comment before the loops

File: pixel_qa/pixel_qa_dilation.c (hunk 7 above)
Comment (raydittmeier) [inline pixel_qa/pixel_qa_dilation.c]: fixed

File: pixel_qa/pixel_qa_dilation.c (hunk 6 above)
Comment (raydittmeier) [inline pixel_qa/pixel_qa_dilation.c]: fixed

**Pull request merged**